### PR TITLE
Complementary timeout bugfix for #1057

### PR DESCRIPTION
**Pull request by ethouris**

### CODE REVIEW

File: srtcore/core.cpp [hunk 1]
```diff
@@ -6026,11 +6026,17 @@ int CUDT::receiveMessage(char *data, int len, SRT_MSGCTRL& w_mctrl)
...
                 HLOGC(tslog.Debug,
-                      log << CONID() << "receiveMessage: fall asleep for " << count_microseconds(recv_timeout) << "us lock=" << (&m_RecvLock)
+                      log << CONID() << "receiveMessage: fall asleep up to TS=" << FormatTime(exptime) << " lock=" << (&m_RecvLock)
                           << " cond=" << (&m_RecvDataCond));
 
-                if (!recv_cond.wait_for(recv_timeout))
+                if (!recv_cond.wait_until(exptime))
                 {
                     if (!(m_iRcvTimeOut < 0))
                         timeout = true;
```
Comment (maxsharabayko) [inline srtcore/core.cpp]: Just poked in my eye
`if (!(m_iRcvTimeOut < 0))` -> `if (m_iRcvTimeOut >= 0)`

Comment (ethouris) [reply srtcore/core.cpp]: right :) I just left the original code here.

Comment (maxsharabayko) [reply srtcore/core.cpp]: Maybe even `recv_timeout` should be used here instead.

**Comment by maxsharabayko** [comment]
Might be the `timeout` variable is also extra for the inner loop.
What if this way? Just break the inner loop if timeout. Then the internal waiting loop can even be moved to a separate small function returning the interrupt reason... 🤔 
```
int  res     = 0;
bool timeout = false;
// Do not block forever, check connection status each 1 sec.
const steady_clock::duration recv_timeout = m_iRcvTimeOut < 0 ? seconds_from(1) : milliseconds_from(m_iRcvTimeOut);

CSync recv_cond (m_RecvDataCond, recvguard);

do
{
    if (stillConnected() && (!m_pRcvBuffer->isRcvDataReady()))
    {
        /* Kick TsbPd thread to schedule next wakeup (if running) */
        if (m_bTsbPd)
        {
            HLOGC(tslog.Debug, log << "receiveMessage: KICK tsbpd");
            tscond.signal_locked(recvguard);
        }

        do
        {
            HLOGC(tslog.Debug,
                  log << CONID() << "receiveMessage: fall asleep for " << count_microseconds(recv_timeout) << "us lock=" << (&m_RecvLock)
                      << " cond=" << (&m_RecvDataCond));

            if (!recv_cond.wait_for(recv_timeout))
            {
                HLOGP(tslog.Debug,
                      "receiveMessage: DATA COND: EXPIRED -- checking connection conditions and rolling again");
                if (m_iRcvTimeOut >= 0)
                {
                    timeout = true;
                    break;
                }
            }
            else
            {
                HLOGP(tslog.Debug, "receiveMessage: DATA COND: KICKED.");
            }
        } while (stillConnected() && (!m_pRcvBuffer->isRcvDataReady()));

        HLOGC(tslog.Debug,
              log << CONID() << "receiveMessage: lock-waiting loop exited: stillConntected=" << stillConnected()
                  << " timeout=" << timeout << " data-ready=" << m_pRcvBuffer->isRcvDataReady());
    }
    
// ...
} while ((res == 0) && !timeout);
```

File: srtcore/core.cpp (hunk 1 above)
Comment (ethouris) [reply srtcore/core.cpp]: No. This is because recv_timeout is set to 1s in case when m_iRcvTimeOut == -1 (no timeout set).

**Pull request merged**